core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/DefaultShadowVariableSessionFactory.java

```diff
@@ -0,0 +1,185 @@
+        var variableIdToUpdater = new HashMap<VariableId, VariableUpdaterInfo>();
+
+        // Maps a variable id to it source aliases;
+        // For instance, "previousVisit.startTime" is a source alias of "startTime"
+        // One way to view this concept is "previousVisit.startTime" is a pointer
+        // to "startTime" of some visit, and thus alias it.
+        Map<VariableId, Set<VariableSourceReference>> declarativeShadowVariableToAliasMap = new HashMap<>();
+
+        // Create graph node for each entity/declarative shadow variable pair
+        for (var entity : entities) {
+            for (var declarativeShadowVariableDescriptor : declarativeShadowVariableDescriptors) {
+                var entityClass = declarativeShadowVariableDescriptor.getEntityDescriptor().getEntityClass();
+                if (entityClass.isInstance(entity)) {
+                    var variableId = new VariableId(entityClass, declarativeShadowVariableDescriptor.getVariableName());
+                    var updater = variableIdToUpdater.computeIfAbsent(variableId, ignored -> new VariableUpdaterInfo(
+                            entityClass,
+                            declarativeShadowVariableDescriptor.getVariableName(),
+                            declarativeShadowVariableDescriptor,
+                            declarativeShadowVariableDescriptor.getEntityDescriptor().getInvalidityMarkerVariableDescriptor(),
+                            declarativeShadowVariableDescriptor.getMemberAccessor(),
+                            declarativeShadowVariableDescriptor.getCalculator()::executeGetter));
+                    variableReferenceGraph.addVariableReferenceEntity(
+                            variableId,
+                            entity,
+                            updater);
+                    for (var sourceRoot : declarativeShadowVariableDescriptor.getSources()) {
+                        for (var source : sourceRoot.variableSourceReferences()) {
+                            if (source.downstreamDeclarativeVariable() != null) {
+                                declarativeShadowVariableToAliasMap
+                                        .computeIfAbsent(source.downstreamDeclarativeVariable(),
+                                                ignored -> new LinkedHashSet<>())
+                                        .add(source);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // Create variable listeners for each declarative shadow variable descriptor
+        for (var declarativeShadowVariable : declarativeShadowVariableDescriptors) {
+            var fromEntityClass = declarativeShadowVariable.getEntityDescriptor().getEntityClass();
+            var fromVariableName = declarativeShadowVariable.getVariableName();
+            final var fromVariableId = new VariableId(fromEntityClass, fromVariableName);
```

Comment:
VariableMetaModel was invented earlier to serve this purpose.
(VariableDescriptor is more of a utility class, and it is a mess. VariableMetaModel is a well-defined carrier of metadata.)